### PR TITLE
Feature generic choice

### PR DESCRIPTION
**Pull request by pmqtt**
Implementation of generic choices!

I hope it is useful!

### CODE REVIEW

File: explorer/interpreter/value.h [hunk 1]
```diff
@@ -838,7 +838,26 @@ class ChoiceType : public Value {
   ChoiceType(std::string name, std::vector<NamedValue> alternatives)
       : Value(Kind::ChoiceType),
         name_(std::move(name)),
-        alternatives_(std::move(alternatives)) {}
+        alternatives_(std::move(alternatives)),
+        is_generic_(false) {}
```
Comment (kmz42) [inline explorer/interpreter/value.h]: Do we need `is_generic_`? Isn't it implied by the presence or absence of bindings (as with `NominalClassType`)?

File: explorer/interpreter/value.h [hunk 2]
```diff
@@ -838,7 +838,26 @@ class ChoiceType : public Value {
...
+
+  ChoiceType(std::string name, std::vector<NamedValue> alternatives,
```
Comment (kmz42) [inline explorer/interpreter/value.h]: maybe i'm missing context but why do we even have a `ChoiceType` constructor that uses `std::string name` instead of only a `ChoiceDeclaration`?

It looks like the pattern we're  using in existing choice code is `arena_->New<ChoiceType>(choice->name(), ...` to construct with the name instead of the `ChoiceDeclaration`. Is there a reason we shouldn't always construct with the `ChoiceDeclaration`?

Comment (kmz42) [reply explorer/interpreter/value.h]: Also, one more thing, could you add comments to each constructor briefly describing the use case for that constructor?

File: explorer/testdata/choice/generic_choice1.carbon [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
```
Comment (kmz42) [inline explorer/testdata/choice/generic_choice1.carbon]: Perhaps add a few more tests, particularly negative tests?

There's a bunch for generic class tests in `testdata/generic_class`. Probably a few of those could become good choice tests.

**Review by kmz42 — COMMENTED** [review]
Added a few thoughts. Still new to the project so feel free to tell me i'm wrong. Nice PR 🎉

File: explorer/interpreter/value.h (hunk 1 above)
Comment (pmqtt) [reply explorer/interpreter/value.h]: Yes, we need is_generic! NominalClassType can check over type_args that binding is present or not.

File: explorer/interpreter/value.h (hunk 2 above)
Comment (pmqtt) [reply explorer/interpreter/value.h]: 1. I try to minimalize code changes! This is the reason why I have not refector this part. But you are right, this should the next step after merging this PR.

File: explorer/interpreter/value.h [hunk 4]
```diff
@@ -851,9 +870,18 @@ class ChoiceType : public Value {
   auto FindAlternative(std::string_view name) const
       -> std::optional<Nonnull<const Value*>>;
 
+  auto bindings() const -> const Bindings& { return *bindings_; }
+  auto type_args() const -> const BindingMap& { return bindings_->args(); }
+  auto declaration() const -> const ChoiceDeclaration& { return *declaration_; }
+
+  auto is_generic() const -> bool { return is_generic_; }
```
Comment (geoffromer) [inline explorer/interpreter/value.h]: For consistency with `NominalClassType`, it seems like this should be called `IsParameterized`, and the return value should be computed from `declaration_` and `bindings_` rather than stored in a separate data member.

File: explorer/interpreter/value.h (hunk 2 above)
Comment (geoffromer) [reply explorer/interpreter/value.h]: > But you are right, this should the next step after merging this PR.

I think it would be simpler to make that change first. Several of the issues I've flagged in other comments would be easier to address that way, and it keeps the API more consistent.

File: explorer/interpreter/value.h [hunk 5]
```diff
@@ -851,9 +870,18 @@ class ChoiceType : public Value {
   auto FindAlternative(std::string_view name) const
       -> std::optional<Nonnull<const Value*>>;
 
+  auto bindings() const -> const Bindings& { return *bindings_; }
+  auto type_args() const -> const BindingMap& { return bindings_->args(); }
+  auto declaration() const -> const ChoiceDeclaration& { return *declaration_; }
```
Comment (geoffromer) [inline explorer/interpreter/value.h]: It looks like it's not safe to call these methods if the object was created using the first constructor.

File: explorer/interpreter/value.h [hunk 6]
```diff
@@ -893,16 +921,25 @@ class ParameterizedEntityName : public Value {
...
   }
 
   auto declaration() const -> const Declaration& { return *declaration_; }
   auto params() const -> const TuplePattern& { return *params_; }
+  auto alternatives() const -> std::vector<NamedValue> { return alternatives_; }
 
  private:
   Nonnull<const Declaration*> declaration_;
   Nonnull<const TuplePattern*> params_;
+  std::vector<NamedValue> alternatives_;
```
Comment (geoffromer) [inline explorer/interpreter/value.h]: If I understand correctly, `ParameterizedEntityName` represents a _name_. In other words, in an expression like `Optional(i32)`, it represents the `Optional` part. That being the case, it seems really weird for this class to have a list of alternatives as a data member. Could the code that needs the alternatives get them out of `declaration` instead?

File: explorer/testdata/choice/generic_choice1.carbon (hunk 3 above)
Comment (geoffromer) [inline explorer/testdata/choice/generic_choice1.carbon]: "generic_choice1" isn't a very descriptive file name. What is this file actually testing that isn't covered by `generic_choice.carbon`?

**Review by geoffromer — DISMISSED** [review]
Thanks, this looks really helpful!

**Comment by pmqtt** [comment]
Hello folks, geoffromer and kmz42,
first thanks for the review. I have rename the test files and prettify the code ( reduceing the constructor count).
Second, I know that store alternatives in ParameterizedType is weird. The reason is, after the type checking process the Interpreter process need this information. The ChoiceDeclaration stores AlternativeSignatures and the ChoiceType need NamedValues. Without a futher type checking step isn't possible to get the NamedValue. Possible solutions are
1. Allow ChoiceDeclaration to store NamedValues 
2. Change the interpreter that generic choices not handle by CallFunction
3. Renaming the ParametrizedType to....?

Another point is, if we want to create the NamedValue in the type checking process TypeCheckPattern and TypeCheckExp we need const_cast. The resulting code is really weird and ugly.

@geoffromer have you any ideas, or should we start with my solution?

**Comment by pmqtt** [comment]
@geoffromer I must apologize... I had thought NamedValue was derived from Value. This is not true, so it is possible to store the Choice-alternatives in the ChoiceDeclaration object.

File: explorer/interpreter/value.h [hunk 7]
```diff
@@ -835,10 +835,13 @@ class SymbolicWitness : public Witness {
...
 class ChoiceType : public Value {
  public:
-  ChoiceType(std::string name, std::vector<NamedValue> alternatives)
+  ChoiceType(Nonnull<const ChoiceDeclaration*> declaration,
+             Nonnull<const Bindings*> bindings)
       : Value(Kind::ChoiceType),
-        name_(std::move(name)),
-        alternatives_(std::move(alternatives)) {}
+        name_(declaration->name()),
+        alternatives_(declaration->members()),
```
Comment (jonmeow) [inline explorer/interpreter/value.h]: Are these copies needed? Since you've started keeping a ChoiceDeclaration for IsParameterized, should name() and FindAlternative use it directly too?

File: explorer/interpreter/type_checker.cpp [hunk 8]
```diff
@@ -2205,8 +2215,11 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
     }
     case ExpressionKind::CallExpression: {
       auto& call = cast<CallExpression>(*e);
+
```
Comment (jonmeow) [inline explorer/interpreter/type_checker.cpp]: I think you're adding a lot of blank lines in this file around code that you're not otherwise editing. The rule of thumb is to minimize vertical whitespace (https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace), and while I think it can be useful to separate code blocks, I think separating out single lines is probably a little more than is needed. Can you please do a pass reducing the whitespace back down?

File: explorer/interpreter/type_checker.cpp [hunk 9]
```diff
@@ -3802,21 +3842,51 @@ auto TypeChecker::TypeCheckImplDeclaration(Nonnull<ImplDeclaration*> impl_decl,
...
   SetConstantValue(choice, ct);
   choice->set_static_type(arena_->New<TypeOfChoiceType>(ct));
   return Success();
 }
 
-auto TypeChecker::TypeCheckChoiceDeclaration(
-    Nonnull<ChoiceDeclaration*> /*choice*/, const ImplScope& /*impl_scope*/)
+auto TypeChecker::TypeCheckChoiceDeclaration(Nonnull<ChoiceDeclaration*> choice,
+                                             const ImplScope& impl_scope)
```
Comment (jonmeow) [inline explorer/interpreter/type_checker.cpp]: These were commented out to indicate they were unused. They probably should not be uncommented in this PR, because uses are not added.

File: explorer/interpreter/resolve_names.cpp [hunk 10]
```diff
@@ -562,14 +562,21 @@ static auto ResolveNames(Declaration& declaration, StaticScope& enclosing_scope,
     }
     case DeclarationKind::ChoiceDeclaration: {
       auto& choice = cast<ChoiceDeclaration>(declaration);
+      StaticScope choice_scope;
+      choice_scope.AddParent(&enclosing_scope);
       enclosing_scope.MarkDeclared(choice.name());
       // Alternative names are never used unqualified, so we don't need to
       // add the alternatives to a scope, or introduce a new scope; we only
       // need to check for duplicates.
```
Comment (jonmeow) [inline explorer/interpreter/resolve_names.cpp]: This comment seems to be referring to the below for loop, and probably shouldn't be separated from it, particularly with a blank line.

File: explorer/interpreter/type_checker.cpp [hunk 11]
```diff
@@ -2233,9 +2247,11 @@ auto TypeChecker::TypeCheckExp(Nonnull<Expression*> e,
...
           // interface to a set of arguments, such as Point(i32) or
           // AddWith(i32).
-          const ParameterizedEntityName& param_name =
-              cast<TypeOfParameterizedEntityName>(call.function().static_type())
-                  .name();
+          const TypeOfParameterizedEntityName& entity =
+              cast<TypeOfParameterizedEntityName>(
+                  call.function().static_type());
+
+          const ParameterizedEntityName& param_name = entity.name();
```
Comment (jonmeow) [inline explorer/interpreter/type_checker.cpp]: This feels like it's not necessary to change. I'm not seeing a particular style benefit to it, and you don't make use of `entity`. Can you please switch back so that we're not changing code when not necessary?

**Review by jonmeow — COMMENTED** [review]
To note on style˛I think there are always going to be differences on preferences for what style is used. While I'm happy to point out style guide issues, I also think it can be beneficial to avoid changing code when there's not a correctness or readability problem. Otherwise, someone else could end up sending a PR to change it back with equivalent justification.

**Comment by jonmeow** [comment]
** picking up this review since geoffromer is out

**Review by jonmeow — APPROVED** [review]
Thanks! I'm happy to have this in.
